Sources/Sentry/SentrySessionTracker.m

```diff
@@ -30,48 +30,54 @@ - (instancetype)initWithOptions:(SentryOptions *)options
+
 #if SENTRY_HAS_UIKIT
-    NSNotificationName foregroundNotificationName = UIApplicationDidBecomeActiveNotification;
-    NSNotificationName backgroundNotificationName = UIApplicationWillResignActiveNotification;
-    NSNotificationName willTerminateNotification = UIApplicationWillTerminateNotification;
+    NSNotificationName didBecomeActiveNotificationName = UIApplicationDidBecomeActiveNotification;
+    NSNotificationName willResignActiveNotificationName = UIApplicationWillResignActiveNotification;
+    NSNotificationName willTerminateNotificationName = UIApplicationWillTerminateNotification;
 #elif TARGET_OS_OSX || TARGET_OS_MACCATALYST
-    NSNotificationName foregroundNotificationName = NSApplicationDidBecomeActiveNotification;
-    NSNotificationName backgroundNotificationName = NSApplicationWillResignActiveNotification;
-    NSNotificationName willTerminateNotification = NSApplicationWillTerminateNotification;
+    NSNotificationName didBecomeActiveNotificationName = NSApplicationDidBecomeActiveNotification;
+    NSNotificationName willResignActiveNotificationName = NSApplicationWillResignActiveNotification;
+    NSNotificationName willTerminateNotificationName = NSApplicationWillTerminateNotification;
 #else
     [SentryLog logWithMessage:@"NO UIKit -> SentrySessionTracker will not "
                               @"track sessions automatically."
                      andLevel:kSentryLogLevelDebug];
 #endif
 
 #if SENTRY_HAS_UIKIT || TARGET_OS_OSX || TARGET_OS_MACCATALYST
-    SentryHub *hub = [SentrySDK currentHub];
-    NSDate *_Nullable lastInForeground =
-        [[[hub getClient] fileManager] readTimestampLastInForeground];
-    if (nil != lastInForeground) {
-        [[[hub getClient] fileManager] deleteTimestampLastInForeground];
-    }
-
-    [hub closeCachedSessionWithTimestamp:lastInForeground];
-    [hub startSession];
 
     [NSNotificationCenter.defaultCenter addObserver:self
                                            selector:@selector(didBecomeActive)
-                                               name:foregroundNotificationName
+                                               name:didBecomeActiveNotificationName
                                              object:nil];
 
     [NSNotificationCenter.defaultCenter addObserver:self
                                            selector:@selector(willResignActive)
-                                               name:backgroundNotificationName
+                                               name:willResignActiveNotificationName
                                              object:nil];
 
     [NSNotificationCenter.defaultCenter addObserver:self
                                            selector:@selector(willTerminate)
-                                               name:willTerminateNotification
+                                               name:willTerminateNotificationName
                                              object:nil];
 
+    [self endCachedSession];
```

Comment:
H: Isn't the order now broken? Since we subscribe to the event before we run this, could the event trigger and race through us accessing the cached session?

Reply:
Very good catch 🥇 . I changed it, but I can't think of a proper way to test it without introducing an adapter around `NSNotificationCenter`. Would you say it's worth the effort? Another advantage would be that we can hide `SENTRY_HAS_UIKIT` and such in there.

Reply:
Sounds good. I think the fix for now is enough.
IMHO we could just merge this and move on, but if you want to add the adapter and test, your call.